msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/AnonymousManagerImpl.java

```diff
@@ -0,0 +1,198 @@
+	}
+
+	public Map<String, String> getUserIdAnonIdMap(final String siteId)
+	{
+		Map<String, String> anonIdMap = new HashMap<String, String>();
+
+		HibernateCallback<List<AnonymousMapping>> hcb = new HibernateCallback<List<AnonymousMapping>>()
+		{
+			public List<AnonymousMapping> doInHibernate(Session session) throws HibernateException, SQLException
+			{
+				Query q = session.getNamedQuery(QUERY_BY_SITE);
+				q.setParameter("siteId", siteId, Hibernate.STRING);
+				return q.list();
+			}
+		};
+
+		List<AnonymousMapping> mappings = getHibernateTemplate().execute(hcb);
+
+		for (AnonymousMapping mapping : mappings)
+		{
+			anonIdMap.put(mapping.getUserId(), mapping.getAnonId());
+		}
+
+		return anonIdMap;
+	}
+
+	public Map<String, String> getUserIdAnonIdMap(final String siteId, final List<String> userIds)
+	{
+		Map<String, String> anonIdMap = new HashMap<String, String>(userIds.size());
+
+		List<AnonymousMapping> mappings = findMappingsBySiteAndUsers(siteId, userIds);
+		if (mappings == null || mappings.isEmpty())
+		{
+			return anonIdMap;
+		}
+
+		for (AnonymousMapping mapping : mappings)
+		{
+			anonIdMap.put(mapping.getUserId(), mapping.getAnonId());
+		}
+
+		return anonIdMap;
+	}
+
+	public Map<String, String> getOrCreateUserIdAnonIdMap(final String siteId, final List<String> userIds)
+	{
+		// Get all the userId - anonId mappings for the specified users in this site
+		Map<String, String> anonIdMap = getUserIdAnonIdMap(siteId, userIds);
+
+		if (anonIdMap.size() < userIds.size())
+		{
+			// The map doesn't contain all of the users. We'll have to create their anonIds
+			Set<String> notFound = new HashSet<String>();
+			notFound.addAll(userIds);
+			notFound.removeAll(anonIdMap.keySet());
+
+			/*
+			 * A race condition could occur:
+			 * 1) request1 gets anonIdMap
+			 * 2) request2 gets anonIdMap
+			 * 3) request1 proceeds to save AnonymousMappings for users in 'notFound'
+			 * 4) request2 tries to save AnonymousMappings for users in 'notFound', but they were created by request 1. Gets a constraint violation
+			 *
+			 * When running on multiple nodes, there is no way around this. 
+			 * But in case we have constraint violations (Ie. we are request 2), we'll add all the users who got constraint violations to this list, and we'll query for their newly created anonIds
+			 */
+			List<String> constraintViolationUsers = new ArrayList<String>();
+			for (String userId : notFound)
+			{
+				String anonId = createAnonId();
+
+				// insert to DB, then put in anonIdMap
+				try
+				{
+					saveAnonMapping(new AnonymousMappingImpl(siteId, userId, anonId));
+					anonIdMap.put(userId, anonId);
+				}
+				catch (DataIntegrityViolationException e)
+				{
+					if (e.getCause() instanceof ConstraintViolationException)
+					{
+						// Race condition!
+						LOG.warn("getOrCreateUserIdAnonIdMap: constraint violation while creating anonId for the following siteId, userId pair: (" + siteId + ", "  + userId + ")");
```

Comment:
As you say, constraint violations are unavoidable here, I would drop this to info because there really isn't any action that can be taken here by anyone reading the logs.